### DISCUSSION TITLE
Usage of the `gix-testtools` crate

### DISCUSSION THREAD

**Discussion by blinxen** [GitoxideLabs/gitoxide#900] [Q&A] [answered]
I can see that the `gix-testtools` crate is published on crates.io but it is never used in the other `gix` crates. The `gix` crates specify it as a `path dependency`. Is that intentional? I am asking this because I am packaging `gix` for fedora and some crates fail to build because their tests are failing. I could simply deactivate the tests, which will fix the build, but I dislike this option.

**Comment by Byron** [comment]
`git-testtools` is used as a `path` dev-dependency most of the time, and that is absolutely intentional as specifying a version would cause trouble when publishing, if that's what you mean.

Tests should be self-contained and are expected to work, but depending on which source for the crates you use, the tests might not be included (like the source packages on crates.io).

If you think tests are failing because they are not isolated enough, I would appreciate an issue with more details. Otherwise I'd also like to figure out why tests are failing, even though my guess is that the test-scripts aren't contained in the cargo packages as fedora probably doesn't build from the git repository.

**Comment by blinxen** [comment]
Thanks for transferring the issue to a discussion! Forgot that GitHub also had this feature :D.

> my guess is that the test-scripts aren't contained in the cargo packages as fedora probably doesn't build from the git repository.

Correct!

The issue that I am experiencing is the following:

```
   Compiling gix-actor v0.21.0 (/builddir/build/BUILD/gix-actor-0.21.0)
     Running `/usr/bin/rustc --crate-name gix_actor --edition=2021 src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --emit=dep-info,link -C opt-level=3 -C embed-bitcode=no --test -C metadata=ab407813c51be34f -C extra-filename=-ab407813c51be34f --out-dir /builddir/build/BUILD/gix-actor-0.21.0/target/release/deps -L dependency=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps --extern bstr=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libbstr-e4b2af27c142c764.rlib --extern btoi=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libbtoi-451b43de909dae80.rlib --extern gix_date=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libgix_date-6d37cc7822acfd84.rlib --extern itoa=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libitoa-fa4dedb26f49e5de.rlib --extern nom=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libnom-ca2c0d1eb54d9508.rlib --extern pretty_assertions=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libpretty_assertions-cb1bf56f19f1ee9f.rlib --extern thiserror=/builddir/build/BUILD/gix-actor-0.21.0/target/release/deps/libthiserror-1ce6cd9357ccf487.rlib -Copt-level=3 -Cdebuginfo=2 -Ccodegen-units=1 -Cforce-frame-pointers=yes -Clink-arg=-Wl,-z,relro -Clink-arg=-Wl,-z,now -Clink-arg=-specs=/usr/lib/rpm/redhat/redhat-package-notes --cap-lints=warn`
error[E0432]: unresolved import `gix_testtools`
  --> src/signature/decode.rs:96:13
   |
96 |         use gix_testtools::to_bstr_err;
   |             ^^^^^^^^^^^^^ use of undeclared crate or module `gix_testtools`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `gix-actor` (lib test) due to previous error
```

The problem here is that the `gix_testtools` module could not be found in the build environment. This is probably because `path dependencies` are not published to `crates.io`, which is used to fetch the sources of crates.  This has the effect that while the crate itself can be built, the tests cannot be run because they are missing the `gix-testtools` crate that is declared as a `path dependency`.

Maybe I should mention that fedora packages all crates as separate packages. That means each crate is equal to one package. When building a package, the build system does not have internet access and relies on dependencies being installed through RPM packages. This is why we encounter the problem from above. I hope I explained this clear enough and did not confuse you more :D.

**Reply by Byron** [reply, marked as answer]
Thanks for elaborating, I definitely am less confused now :).

`gix-testtools` has been published to crates.io so a package could be created for it. But to pick it up, I presume the cargo-manifests would need to be patched.

Further, since the `gix` crate packages don't contain any test code, there isn't much to run in the first place. The few tests that are included as unit-tests aren't that important so skipping them isn't a big deal. Maybe that's the only way forward then?

**Reply by blinxen** [reply]
> The few tests that are included as unit-tests aren't that important so skipping them isn't a big deal. Maybe that's the only way 
forward then?

Yeah, I think that is what I am going to do. If something changes in the future, reactivating the tests should be trivial.

> But to pick it up, I presume the cargo-manifests would need to be patched

This could be another option but I think this requires too much effort for very little gain.

**Comment by blinxen** [comment]
@Byron With the [latest release of `gix-testtools`](https://github.com/GitoxideLabs/gitoxide/releases/tag/gix-testtools-v0.19.0), would it make sense to now package it so we can run gitoxide tests in Fedora?

**Reply by Byron** [reply]
Yes, I think that would work. It shouldn't be special anymore. Please let me know what you find if it's not working though.
